### DISCUSSION TITLE
Supporting Versioned tables

### DISCUSSION THREAD

**Discussion by jasonamyers** [sqlalchemy/sqlalchemy#11161] [Usage Questions]
Some table types like Iceberg, Hudi, etc support querying the table at a particular version or point in time. For example:

`SELECT * FROM iceberg_table FOR VERSION AS OF 949530903748831860`
`SELECT * FROM iceberg_table FOR TIMESTAMP AS OF TIMESTAMP '2020-01-01 10:00:00 UTC'`

I'm figuring out how to properly approach these version clauses in a dialect. I think I need to alter the default_from and perhaps the table metadata to support having a version number or timestamp, but I'm really unsure the proper course of action here.

**Comment by zzzeek** [comment]
this is called temporal queries and there's a whole issue / PR for this that currently got closed at #10557, since the PR was very complicated and we have never had anyone else other than that one contributor ever request it, and they went away, likely because I was having trouble being patient for that one.

so your ask here is basically the **second** time someone is looking for it, which possibly changes everything , in that OK maybe we should try to figure this one out.

as I mentioned in my comment at https://github.com/sqlalchemy/sqlalchemy/discussions/10557#discussioncomment-7411958 there is probably a way to build custom constructs for this.    One complexity seems to be that there's firstly the whole DDL side of it, then there's the whole SELECT/DQL side of it.   the above PR was stuck just on DDL.

as far as specifics for SELECT, I think my comment in https://github.com/sqlalchemy/sqlalchemy/issues/4797#issuecomment-798451209 is closest to how this might look.   this is essentially a modifier for a table in the FROM clause, and my experience with things like LATERAL etc. have shown we might as well make a custom FROM clause container that carries along the special directives that can render in the FROM clause.

the reason the whole feature didnt move forward is that we'd want our custom FROM clause to integrate fully with the DDL constructs, and there are too many moving parts for me to make good decisions without fully immersing myself in these syntaxes.  However, that you're pointing out a few more DBs that do this is gerat, because usually we want to make a matrix of all the DBs that support "the thing" and make sure we cover all the cases.

cc @CaselIT

**Reply by jasonamyers** [reply]
Understood, we see more and more analytics, lakehouse-type engines, being used to fix the data at a version/tag that has been QAed, use the data beyond that tag in the QA process, and advance the version forward in queries when data is approved. At least in the Athena and Snowflake usages of the Iceberg/Hudi table formats, you can't do it with DDL pieces; it's just a part of the FROM causes in queries.

**Reply by jasonamyers** [reply]
Reading through that other thread, I feel like I could implement some with_hint handling in the Pyathena dialect to add it. Does that feel appropriate to you as a starting point for this?

**Reply by zzzeek** [reply]
i think it does.   keep me in the loop what you come up with

**Reply by jasonamyers** [reply]
This is the way I ended up going.  The alias part feels messy, but I don't think I have access to what I need there to get around it.

*Changes to Athena Compiler*
```python
class AthenaStatementCompiler(SQLCompiler):

    def get_from_hint_text(self, table, text):
        """ Parses the text of a FROM clause hint, returning the text formatted properly"""
        return text

    def format_from_hint_text(self, sqltext, table, hint, iscrud):
        hint_upper = hint.upper()

        if not any([
            hint_upper.startswith("FOR TIMESTAMP AS OF"),
            hint_upper.startswith("FOR SYSTEM_TIME AS OF"),
            hint_upper.startswith("FOR VERSION AS OF"),
            hint_upper.startswith("FOR SYSTEM_VERSION AS OF"),
        ]):
            raise exc.CompileError("Unrecognized hint: %r" % hint)

        if 'AS' in sqltext.upper():
            _, alias = sqltext.upper().split(' AS ', 1)
            return f"{table.original.fullname} {hint} AS {alias.lower()}"

        return  f"{sqltext} {hint}"
```

*Usage example*
```python
engine = get_engine('athena')
table = build_table(engine, 'my_table', 'my_schema')
ta = table.alias()
version = 8860342768238431084
query = select(func.count(ta.c.record_id)).with_hint(ta, f'FOR VERSION AS OF {version}')
result = str(query.compile(compile_kwargs={"literal_binds": True}, dialect=AthenaDialect()))
```

*Outcome*
```output
SELECT count(my_table.record_id) AS count_1 
FROM my_schema.my_table FOR VERSION AS OF 8860342768238431084 AS my_table_1
```

**Reply by zzzeek** [reply]
I dont think it's too terrible.   Why not use your startswith as the conditonal to process the text, and let others pass?  like this:


```py
class AthenaStatementCompiler(SQLCompiler):

    def get_from_hint_text(self, table, text):
        """ Parses the text of a FROM clause hint, returning the text formatted properly"""
        return text

    def format_from_hint_text(self, sqltext, table, hint, iscrud):
        hint_upper = hint.upper()

        if any([
            hint_upper.startswith("FOR TIMESTAMP AS OF"),
            hint_upper.startswith("FOR SYSTEM_TIME AS OF"),
            hint_upper.startswith("FOR VERSION AS OF"),
            hint_upper.startswith("FOR SYSTEM_VERSION AS OF"),
        ]):
            _, alias = sqltext.upper().split(' AS ', 1)
            return f"{table.original.fullname} {hint} AS {alias.lower()}"

        return  f"{sqltext} {hint}"
```

**Reply by jasonamyers** [reply]
Awesome here is the PR where we're working on it. https://github.com/laughingman7743/PyAthena/pull/518
